icarusalg/Utilities/AssnsCrosser.h

```diff
@@ -0,0 +1,1749 @@
+        (event, std::move(otherInputSpecs)... );
+    default:
+      throw std::logic_error
+        { "Unexpected direction: " + std::to_string(static_cast<int>(algo)) };
+  } // switch
+} // icarus::ns::util::AssnsCrosser<>::prepare()
+
+
+// -----------------------------------------------------------------------------
+template <typename KeyType, typename... OtherTypes>
+auto icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>
+  ::chooseTraversalAlgorithm
+(
+  StartSpecs<KeyType> const& startSpecs,
+  InputSpecs<OtherTypes> const&... otherInputSpecs
+) const -> HoppingAlgo {
+  /*
+   * If there is a start specification, we need (and can) to go forward.
+   * Otherwise, we go backward (faster) unless there is no specification for
+   * the last hop (in which case we can't start from the back).
+   * 
+   * When both algorithms are available, we prefer the backward one.
+   */
+  
+  bool const hasStartInfo = startSpecs.hasSpecs();
+  
+  bool const hasEndSpecs
+    = details::getElement<-1>(otherInputSpecs...).hasSpecs();
+  
+  constexpr std::size_t nHops = sizeof...(OtherTypes);
+  
+#if 0
+  // --- BEGIN -- DEBUG --------------------------------------------------------
+  // print details about how the specifications are received:
+  auto const formatter
+    = [&out=std::cout](auto const& specs){ out << "\n  " << specs; };
+  std::cout
+    << "Start specs: " << startSpecs << ", hasStartInfo=" << hasStartInfo
+    << "\n" << nHops << " hops:";
+  (formatter(otherInputSpecs), ...);
+  std::cout << "\nLast spec: " << details::getElement<-1>(otherInputSpecs...)
+    << " -> hasEndSpecs=" << hasEndSpecs << std::endl;
+  
+  // --- END ---- DEBUG --------------------------------------------------------
+#endif // 0
+  if constexpr(nHops == 1) {
+    if (hasStartInfo) return HoppingAlgo::forward;
+    if (hasEndSpecs) return HoppingAlgo::backward;
+    throw std::logic_error
+      { "Insufficient specifications for single association traversal." };
+  }
+  else {
+    bool const hasFirstSpecs
+      = hasStartInfo || details::getElement<0>(otherInputSpecs...).hasSpecs();
+    
+    if (hasFirstSpecs) return HoppingAlgo::forward;
+    if (hasEndSpecs) return HoppingAlgo::backward;
+    
+    throw std::logic_error{
+      "Insufficient specifications for traversal of " + std::to_string(nHops)
+      + " associations."
+      };
+    
+  }
+  
+} // icarus::ns::util::AssnsCrosser<>::chooseTraversalAlgorithm()
+
+
+// -----------------------------------------------------------------------------
+template <typename KeyType, typename... OtherTypes>
+template <typename T, typename Event>
+auto icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>::keysFromSpecs
+  (Event const& event, StartSpecs<T> const& specs) const
+  -> std::optional<details::PointerSelector<T>>
+{
+  /*
+   * An interface needs to be established.
+   * The specification may follow the pattern of the InputSpecs, but but can't
```

Comment:
Extra "but" in this line

Reply:
But... but... but... ok, ok.
Extra ok there.